ports/nrf/mphalport.c

```diff
@@ -35,6 +35,120 @@
+        nrf_clock_task_trigger(NRF_CLOCK, NRF_CLOCK_TASK_LFCLKSTART);
+    }
+    nrfx_rtc_init(&rtc1, &rtc_config_time_ticks, rtc_irq_time);
+    nrfx_rtc_overflow_enable(&rtc1, true);
+    // generate a 1msec irq to leave _WFI
+    nrfx_rtc_cc_set(&rtc1, 0, 30, true);  
+    nrfx_rtc_enable(&rtc1);
+}
+
+mp_uint_t mp_hal_ticks_ms(void) {
+    // calculate (rtc_overflows << 24 + COUNTER) * 1000 / 32768
+    // 
+    // note that COUNTER * 1000 / 32768 would overflow during calculation, so use
+    // the less obvious * 125 / 4096 calculation (overflow secure)
+    //
+    // also make sure not to call this function within an irq with higher
+    // prio than the RTC's irq. This would introduce the danger of 
+    // preempting the RTC irq and calling mp_hal_ticks_ms() at that time
+    // would return a false result
+    //
+    // also lets guard the function against interrupting from the RTC 
+    // overflow irq while the calculation takes place
+
+    // see also mp_hal_ticks_us() for a description of the below code
+    rtc1.p_reg->INTENCLR = RTC_INTENCLR_OVRFLW_Msk;
+    uint32_t overflows = rtc_overflows;
+    uint32_t counter = rtc1.p_reg->COUNTER;
+    uint32_t has_overflowed = rtc1.p_reg->EVENTS_OVRFLW;
+    if (has_overflowed && counter < (1 << 23)) {
+        overflows += 1;
+    }
+    rtc1.p_reg->INTENSET = RTC_INTENSET_OVRFLW_Msk;
+    return (overflows << 9) * 1000 + (counter * 125 / 4096);
+}
+
+mp_uint_t mp_hal_ticks_us(void) {
+    // make the best out of the 32kHz tick resolution (= 30.57usec / tick)
+    // to scale we could use same approach as above and use COUNTER * 125000 / 4096
+    // but that will overflow for 32bit multiplications.
+    //
+    // Since this function is likely to be called in a poll loop it must
+    // be fast and casting to 64 bits adds calculation time within that loop,
+    // causing timing inaccuracy.
+    
+    // state = disable_irq(); // fully disable interrupts
+    rtc1.p_reg->INTENCLR = RTC_INTENCLR_OVRFLW_Msk;
+    uint32_t overflows = rtc_overflows;
+    uint32_t counter = rtc1.p_reg->COUNTER;
+    // note: COUNTER may overflow just after reading it and before reading EVENTS_OVRFLW
+    uint32_t has_overflowed = rtc1.p_reg->EVENTS_OVRFLW;
+    // If has_overflowed is set it could be before or after COUNTER is read. 
+    // If before then an adjustment must be made, if after then no adjustment is necessary.
+    //  
+    // The before case is when COUNTER is very small (because it just overflowed and was set to zero), 
+    // the after case is when COUNTER is very large (because it's just about to overflow 
+    // but we read it right before it overflows). 
+    // 
+    // The extra check for counter is to distinguish these cases. 1<<23 because it's halfway 
+    // between min and max values of COUNTER.
+    if (has_overflowed && counter < (1 << 23)) {
+        overflows += 1;
+    }
+    rtc1.p_reg->INTENSET = RTC_INTENSET_OVRFLW_Msk;
+    // enable_irq(state);
```

Comment:
would be good to factor this chunk out to a macro (eg `RTC1_GET_TICKS_ATOMIC`) and use it here and above in `mp_hal_ticks_ms()`

Reply:
ok done